### ISSUE TITLE
Low WER training pipeline in torchaudio with wav2letter

### ISSUE THREAD

**Issue by vincentqb** [pytorch/audio#913]
torchaudio is targeting speech recognition as full audio application ([internal](https://docs.google.com/document/d/1mygzn6vf6Jl4e_0fzJSciRGRmb8dj8NwibwBG-tyUj8/edit?ts=5ef3580b#bookmark=id.7evu2kt47eiz)). Along this line, we implemented [wav2letter pipeline](https://github.com/pytorch/audio/pull/632) to obtain a low character error rate (CER). We want to expand on this and showcase a new pipeline which also has a low word error rate (WER). To achieve this, we consider the following additions to torchaudio from higher to lower priority.

**Token Decoder:** Add a lexicon-constrained beam search algorithm, based on  fairseq ([search class](https://github.com/pytorch/fairseq/blob/master/fairseq/search.py), [sequence generator](https://github.com/pytorch/fairseq/blob/master/fairseq/sequence_generator.py#L272)) since it is torchscriptable.

* Links: [fairseq](https://github.com/pytorch/fairseq/blob/master/fairseq/search.py), [wav2letter](https://github.com/facebookresearch/wav2letter/tree/v0.2/src/libraries/decoder), [ParlAI](https://github.com/facebookresearch/ParlAI/blob/3d3386caa9a1334a5846ae49b6878e3c28f4990d/parlai/core/torch_generator_agent.py#L940), [user repository](https://github.com/budzianowski/PyTorch-Beam-Search-Decoding), [caffe2](https://caffe2.ai/doxygen-python/html/beam__search_8py_source.html), [pyspeech](https://www.internalfb.com/intern/diffusion/FBS/browse/master/fbcode/deeplearning/projects/pyspeech/pyspeech/fb/external_decoder).
* Kaldi related: Kaldi Viterbi BeamSearch [SimpleDecoder](https://kaldi-asr.org/doc/lattices.html), [internal](https://fb.quip.com/FsPWAFfvufsZ), and [OpenFST](http://www.openfst.org/twiki/bin/view/FST/WebHome).
* Other domains: paper for [vision](http://vision.stanford.edu/cs598_spring07/papers/Lecun98.pdf)
* Related algorithm: FAISS’s topk [internal](https://fb.workplace.com/groups/434225650103564/permalink/1401892803336839/), [paper](https://arxiv.org/pdf/1702.08734.pdf) and [github](https://github.com/facebookresearch/faiss)

**Acoustic Model:** Add a transformer-based acoustic model, e.g. [speech-transformer](http://150.162.46.34:8080/icassp2018/ICASSP18_USB/pdfs/0005884.pdf), [comparison](https://arxiv.org/pdf/1909.06317v2.pdf).

* wav2vec 2.0 ([paper](https://arxiv.org/abs/2006.11477), fairseq’s [github](https://github.com/pytorch/fairseq/blob/master/examples/wav2vec/README.md)) combines transformer and lexicon-constrained beam search
* [multi-speaker](https://arxiv.org/pdf/2002.03921.pdf), streaming ([paper](https://arxiv.org/pdf/2005.08042.pdf), [post](https://fb.workplace.com/groups/831302610278251/permalink/4284333054975172/))

**Language Model:** Add [KenLM](https://github.com/kpu/kenlm) to use a 4-gram language model based on [LibriSpeech Language Model](http://www.openslr.org/11/), as done in [paper](https://arxiv.org/pdf/2005.08042.pdf).

* [wav2letter](https://github.com/facebookresearch/wav2letter/wiki/Beam-Search-Decoder) interfaces with both [KenLM](https://github.com/kpu/kenlm) and [ConvLM](https://arxiv.org/pdf/1812.06864.pdf)
* This could land in torchtext.

**Training Loss:** Add the RNN Transducer loss to replace the CTC loss in the pipeline.

* RNNT ([paper](https://arxiv.org/pdf/1211.3711.pdf), [github](https://github.com/HawkAaron/warp-transducer)) 
* see also [internal](https://fb.quip.com/Emo2AcGAEc8U)

**Transformations:** SpecAugment is already available in wav2letter [pipeline](https://github.com/pytorch/audio/pull/632).

See also [internal](https://fb.quip.com/Emo2AcGAEc8U)

cc @astaff @dongreenberg @cpuhrsch

**Comment by scarecrow1123** [comment]
Hey @vincentqb these planned additions look great and useful! Can you clarify on the below points please? 

1. Are these open for community contributions since there are a bunch of internal links you've linked to the description above?

2. For things like transducer loss and token decoder, are you planning on linking other libraries (like fairseq, warp-transducer, etc.) as dependencies to the example application? Or would they be self contained as it is done for wav2letter implementation in #632 ?

**Comment by vincentqb** [comment]
Hey, thanks for commenting :)

1. For now, we are looking for high level feedback on the plan. This has not yet been taskyfied for the community :)

1. We are aiming at keeping the implementations self-contained in torchaudio as much as possible.

Thoughts?

**Comment by Edresson** [comment]
@vincentqb  Wouldn't it be interesting to create a torchaudio_ASR repository?

Support various things, an independent preprocess of features that saves the features in .pt so we can extract any feature from torchaudio and make an easy support to features from fairseq (like wav2vec).

I feel that ASR lacks a repository with torchaudio and is modular enough to accept new features. And that it is simple for people to add new models (which are different from those implemented in torchaudio (currently we only have wav2letter)). I believe that the community can like this idea and contribute to the repository.

**Comment by vincentqb** [comment]
> @vincentqb Wouldn't it be interesting to create a torchaudio_ASR repository?

My first step is to understand what would be missing in torchaudio to serve the ASR community best :) Can you provide some examples?

> Support various things, an independent preprocess of features that saves the features in .pt so we can extract any feature from torchaudio and make an easy support to features from fairseq (like wav2vec).

`torch.save` can save any `nn.Module` and tensors. Is that what you meant?

> I feel that ASR lacks a repository with torchaudio and is modular enough to accept new features. And that it is simple for people to add new models (which are different from those implemented in torchaudio (currently we only have wav2letter)). I believe that the community can like this idea and contribute to the repository.

Are there models you would like to contribute to torchaudio? :)

**Comment by Edresson** [comment]
@vincentqb  I would support the Jasper model. Right now I'm out of time so maybe soon I can send a PR :).

If that list of things has been added to torchaudio it will be very good for ASR :). It will be even simpler to build a great pipeline using only torchaudio :).

My initial suggestion is to keep external notebooks. Why not make a torchaudio_ASR repository? So some things would not need to be implemented in the torchaudio itself, but in these new repositories.  There in this repository we can extract features before and save with torch.save and just write a generic dataloader that reads, this is interesting to facilitate support for new features like wav2vec. So to support new extraction methods independent of torchaudio, just write a new preprocessing class.

**Comment by vincentqb** [comment]
> @vincentqb I would support the Jasper model. Right now I'm out of time so maybe soon I can send a PR :).

Great! Feel free to ping me when you do :)

> If that list of things has been added to torchaudio it will be very good for ASR :). It will be even simpler to build a great pipeline using only torchaudio :).
> 
> My initial suggestion is to keep external notebooks. 

We currently offer training examples such as [wav2letter](https://github.com/pytorch/audio/tree/master/examples/pipeline_wav2letter) using torchaudio. The example I link shows one way of doing preprocessing. Does that help?

> Why not make a torchaudio_ASR repository? So some things would not need to be implemented in the torchaudio itself, but in these new repositories. There in this repository we can extract features before and save with torch.save and just write a generic dataloader that reads, this is interesting to facilitate support for new features like wav2vec. So to support new extraction methods independent of torchaudio, just write a new preprocessing class.

Our goal with torchaudio is to provide flexible building blocks for audio-related fields, such as ASR. As such, we want to make sure we capture what would be useful to the community, and to ASR. Can you provide an example of your suggested workflow?

**Comment by Edresson** [comment]
> We currently offer training examples such as wav2letter using torchaudio. The example I link shows one way of doing preprocessing. Does that help?

I had not seen this very good example :)

> Our goal with torchaudio is to provide flexible building blocks for audio-related fields, such as ASR. As such, we want to make sure we capture what would be useful to the community, and to ASR. Can you provide an example of your suggested workflow?

Now that I've seen the example. Do you think about supporting [Wav2vec]( https://github.com/pytorch/fairseq/tree/master/examples/wav2vec)?

The easiest way to do this support that I see is to change the example, separating the feature extraction (MFFC / waveform) from the model training. Basically a preprocess.py that extracts the characteristics and saves them with torch.save so the main only reads the files saved by torch.save.

Do you know a simpler way to integrate Wav2vec with torchaudio?

**Comment by vincentqb** [comment]
Adding an example workflow with wav2vec would be a great addition! I see you have already mentioned jasper in [comment](https://github.com/pytorch/audio/issues/446#issuecomment-650634109) so let's move the discussion there :)

> Now that I've seen the example. Do you think about supporting [Wav2vec](https://github.com/pytorch/fairseq/tree/master/examples/wav2vec)?
> 
> The easiest way to do this support that I see is to change the example, separating the feature extraction (MFFC / waveform) from the model training. Basically a preprocess.py that extracts the characteristics and saves them with torch.save so the main only reads the files saved by torch.save.
> 
> Do you know a simpler way to integrate Wav2vec with torchaudio?

We currently don't have a pipeline with wav2vec included, but this would be a great addition. torchaudio is made to be modular and uses standard pytorch operations, so using the pre-trained tensors from fairseq can be done using standard pytorch operations. Is that what you meant?

**Comment by Edresson** [comment]
> Adding an example workflow with wav2vec would be a great addition! I see you have already mentioned jasper in [comment](https://github.com/pytorch/audio/issues/446#issuecomment-650634109) so let's move the discussion there :)
> 
> > Now that I've seen the example. Do you think about supporting [Wav2vec](https://github.com/pytorch/fairseq/tree/master/examples/wav2vec)?
> > The easiest way to do this support that I see is to change the example, separating the feature extraction (MFFC / waveform) from the model training. Basically a preprocess.py that extracts the characteristics and saves them with torch.save so the main only reads the files saved by torch.save.
> > Do you know a simpler way to integrate Wav2vec with torchaudio?
> 
> We currently don't have a pipeline with wav2vec included, but this would be a great addition. torchaudio is made to be modular and uses standard pytorch operations, so using the pre-trained tensors from fairseq can be done using standard pytorch operations. Is that what you meant?

Do you have any idea how this support would do?
Do you intend to add fairseq as a dependency on the torchaudio or example in question?

Or do you want to try to make it independent of the fairseq structure and just create a class with Wav2vec architecture and load it at the checkpoint?

**Comment by vincentqb** [comment]
Those are great questions, and thanks for sharing your thoughts :)

> Do you intend to add fairseq as a dependency on the torchaudio or example in question?
> Or do you want to try to make it independent of the fairseq structure and just create a class with Wav2vec architecture and load it at the checkpoint?

We do not want torchaudio to depend on fairseq, no. For the example implementation, we also aim to avoid such dependencies as much as possible.

I would aim instead for fairseq to use torchaudio building blocks in some places.

> Do you have any idea how this support would do?

What I meant above about the checkpoint was really just that torchaudio uses standard pytorch, so a user can interact with through standard pytorch means. For instance, someone could pre-process with torchaudio on a torchaudio dataset and then import a model from somewhere else, and then follow a torchaudio example for training loop. 

Is this what you meant?

**Comment by Edresson** [comment]
> @Edresson -- those are great questions, and thanks for sharing your thoughts :)
> 
> > Do you intend to add fairseq as a dependency on the torchaudio or example in question?
> > Or do you want to try to make it independent of the fairseq structure and just create a class with Wav2vec architecture and load it at the checkpoint?
> 
> We do not want torchaudio to depend on fairseq, no. For the example implementation, we also aim to avoid such dependencies as much as possible.
> 
> I would aim instead for fairseq to use torchaudio building blocks in some places.
> 
> > Do you have any idea how this support would do?
> 
> What I meant above about the checkpoint was really just that torchaudio uses standard pytorch, so a user can interact with through standard pytorch means. For instance, someone could pre-process with torchaudio on a torchaudio dataset and then import a model from somewhere else, and then follow a torchaudio example for training loop.
> 
> Is this what you meant?

I believe so :). We could, for example, preprocess the dataset with torchaudio after extracting the features of these audios with wav2vec and save with torch.save. After going back to torchaudio and using the torchaudio ASR models. I think of saving with torch.save because extracting features at all times with wav2vec can be too slow, so the extraction is done only once. That makes sense?

**Comment by vincentqb** [comment]
> I believe so :). 

great!

> We could, for example, preprocess the dataset with torchaudio after extracting the features of these audios with wav2vec and save with torch.save. After going back to torchaudio and using the torchaudio ASR models. I think of saving with torch.save because extracting features at all times with wav2vec can be too slow, so the extraction is done only once. That makes sense?

yup, does to me :)
